blueman/plugins/applet/Menu.py

```diff
@@ -24,50 +100,52 @@ class Menu(AppletPlugin):
     def on_load(self, applet):
         self.Applet = applet
 
-        self.Applet.Plugins.StatusIcon.connect("popup-menu", self.on_popup_menu)
-
         self.__plugins_loaded = False
 
         self.__menuitems = []
-        self.__menu = Gtk.Menu()
-
-    def on_popup_menu(self, status_icon, button, activate_time):
-        self.__menu.popup(None, None, Gtk.StatusIcon.position_menu,
-                          status_icon, button, activate_time)
 
     def __sort(self):
-        self.__menuitems.sort(key=itemgetter(0))
-
-    def __clear(self):
-        def each(child, _):
-            self.__menu.remove(child)
-
-        self.__menu.foreach(each, None)
-
-    def __load_items(self):
-        for item in self.__menuitems:
-            self.__menu.append(item[1])
-            if item[2]:
-                item[1].show()
+        self.__menuitems.sort(key=attrgetter('priority'))
 
-    def Register(self, owner, item, priority, show=True):
-        self.__menuitems.append((priority, item, show, owner))
+    def add(self, owner, priority, text=None, markup=False, icon_name=None, tooltip=None, callback=None,
+            submenu_function=None, visible=True, sensitive=True):
+        item = MenuItem(self, owner, priority, text, markup, icon_name, tooltip, callback, submenu_function, visible,
+                        sensitive)
+        self.__menuitems.append(item)
         if self.__plugins_loaded:
             self.__sort()
-            self.__clear()
-            self.__load_items()
+        self.on_menu_changed()
+        return item
 
-    def Unregister(self, owner):
-        for i in reversed(self.__menuitems):
-            priority, item, show, orig_owner = i
-            if orig_owner == owner:
-                self.__menu.remove(item)
-                self.__menuitems.remove(i)
+    def unregister(self, owner):
+        for item in reversed(self.__menuitems):
+            if item.owner == owner:
+                self.__menuitems.remove(item)
+        self.on_menu_changed()
 
     def on_plugins_loaded(self):
         self.__plugins_loaded = True
         self.__sort()
-        self.__load_items()
 
-    def get_menu(self):
-        return self.__menu
+    def on_menu_changed(self):
+        self.MenuChanged(self.GetMenu())
```

Comment:
Why send the whole menu? Would be simpler if the the client (Tray) just calls ``AppletService.GetMenu`` instead?